lib/tests/misc.nix

```diff
@@ -761,4 +761,143 @@ runTests {
       { a = 3; b = 30; c = 300; }
     ];
   };
+
+  testShowAttrPathEmpty = {
+    expr = showAttrPath [];
+    expected = "<root attribute path>";
+  };
+
+  testShowAttrPathVarious = {
+    expr = showAttrPath [
+      "."
+      "foo"
+      "2"
+      "a2-b"
+      "_bc'de"
+    ];
+    expected = "\".\".foo.\"2\".a2-b._bc'de";
+  };
+
+  # If there are no updates, the value is passed through
+  testUpdateManyAttrPathsNone = {
+    expr = updateManyAttrPaths [] "something";
+    expected = "something";
+  };
+
+  # A single update to the root path is just like applying the function directly
+  testUpdateManyAttrPathsSingleIncrement = {
+    expr = updateManyAttrPaths [
+      {
+        path = [ ];
+        update = old: old + 1;
+      }
+    ] 0;
+    expected = 1;
+  };
+
+  # Multiple updates can be applied are done in order
+  testUpdateManyAttrPathsMultipleIncrements = {
+    expr = updateManyAttrPaths [
+      {
+        path = [ ];
+        update = old: old + "a";
+      }
+      {
+        path = [ ];
+        update = old: old + "b";
+      }
+      {
+        path = [ ];
+        update = old: old + "c";
+      }
+    ] "";
+    expected = "abc";
+  };
+
+  # If an update doesn't use the value, all previous updates are not evaluated
+  testUpdateManyAttrPathsLazy = {
+    expr = updateManyAttrPaths [
+      {
+        path = [ ];
+        update = old: old + throw "nope";
+      }
+      {
+        path = [ ];
+        update = old: "untainted";
+      }
+    ] (throw "start");
+    expected = "untainted";
+  };
+
+  # Deeply nested attributes can be updated without affecting others
+  testUpdateManyAttrPathsDeep = {
+    expr = updateManyAttrPaths [
+      {
+        path = [ "a" "b" "c" ];
+        update = old: old + 1;
+      }
+    ] {
+      a.b.c = 0;
+
+      a.b.z = 0;
+      a.y.z = 0;
+      x.y.z = 0;
+    };
+    expected = {
+      a.b.c = 1;
+
+      a.b.z = 0;
+      a.y.z = 0;
+      x.y.z = 0;
+    };
+  };
+
+  # Nested attributes are updated first
+  testUpdateManyAttrPathsNestedBeforehand = {
+    expr = updateManyAttrPaths [
+      {
+        path = [ "a" ];
+        update = old: old // { x = old.b; };
+      }
+      {
+        path = [ "a" "b" ];
+        update = old: old + 1;
+      }
+    ] {
+      a.b = 0;
+    };
+    expected = {
+      a.b = 1;
+      a.x = 1;
+    };
+  };
+
+  # If an attribute doesn't exist, it is treated as if it did with a value of {}
```

Comment:
This is too arbitrary for my taste at least.
How about adding `setWhenMissing` next to `update` to handle this case?

Reply:
The reason I chose `{}` is because this allows easily setting nested attributes:

```
nix-repl> :p lib.updateManyAttrPaths [ { path = [ "a" "b" "c" ]; update = _: 10; } ] {}   
{ a = { b = { c = 10; }; }; }
```

This works because when it tries to recurse into `a` and finds that there's no such attribute, it creates it with a value of `{}`. The same for the next path element.

This wouldn't work if e.g. `null` were used instead in the current implementation, it would result in an error:
```
nix-repl> lib.updateManyAttrPaths [ { path = [ "a" "b" "c" ]; update = old: 10; } ] {} 
{ a = «error: updateManyAttrPaths: Path a.b.c needs to be updated, but path a of the given input is not an attribute set, so we can't update an attribute inside of it.»; }
```

But after thinking about this again, this default of `{}` really only makes sense for initializing these intermediate path elements. For the last path element of an update, I very much like your suggestion of making it configurable. I'll try to implement this.

Reply:
I revisited this and determined that no such functionality is needed, because you can implement a fallback by just inserting another update. E.g. this example with your proposed `setWhenMissing`:

```nix
updateManyAttrPaths [
  {
    path = [ "a" ];
    update = old: old + "-updated";
    setWhenMissing = "missing";
  }
] {}
```

is equivalent to

```nix
updateManyAttrPaths [
  {
    path = [ "a" ];
    update = old: "missing";
  }
  {
    path = [ "a" ];
    update = old: old + "-updated";
  }
] {}
```

Reply:
However, I did remove the arbitrary default of `{}`, as it was only needed internally. Instead you will now get an error if you try to access an `old` argument when it doesn't exist.

Reply:
> ```nix
>     update = old: "missing";
> ```

Wouldn't that be destructive?

> ```nix
>     update = old: old + "-updated";
> ```

Or _if_ it wasn't destructive, wouldn't `old` still reference the exception?